eth2/fork_choice/src/optimized_lmd_ghost.rs

```diff
@@ -0,0 +1,452 @@
+
+            block_slot.height(spec.genesis_slot)
+        };
+
+        // verify we haven't exceeded the block height
+        if target_height >= block_height {
+            if target_height > block_height {
+                return None;
+            } else {
+                return Some(block_hash);
+            }
+        }
+        // check if the result is stored in our cache
+        let cache_key = CacheKey::new(&block_hash, target_height.as_u64());
+        if let Some(ancestor) = self.cache.get(&cache_key) {
+            return Some(*ancestor);
+        }
+
+        // not in the cache recursively search for ancestors using a log-lookup
+        if let Some(ancestor) = {
+            let ancestor_lookup = self.ancestors
+                [log2_int((block_height - target_height - 1u64).as_u64()) as usize]
+                .get(&block_hash)
+                //TODO: Panic if we can't lookup and fork choice fails
+                .expect("All blocks should be added to the ancestor log lookup table");
+            self.get_ancestor(*ancestor_lookup, target_height, &spec)
+        } {
+            // add the result to the cache
+            self.cache.insert(cache_key, ancestor);
+            return Some(ancestor);
+        }
+
+        None
+    }
+
+    // looks for an obvious block winner given the latest votes for a specific height
+    fn get_clear_winner(
+        &mut self,
+        latest_votes: &HashMap<Hash256, u64>,
+        block_height: SlotHeight,
+        spec: &ChainSpec,
+    ) -> Option<Hash256> {
+        // map of vote counts for every hash at this height
+        let mut current_votes: HashMap<Hash256, u64> = HashMap::new();
+        let mut total_vote_count = 0;
+
+        trace!("Clear winner at block height: {}", block_height);
+        // loop through the latest votes and count all votes
+        // these have already been weighted by balance
+        for (hash, votes) in latest_votes.iter() {
+            if let Some(ancestor) = self.get_ancestor(*hash, block_height, spec) {
+                let current_vote_value = current_votes.get(&ancestor).unwrap_or_else(|| &0);
+                current_votes.insert(ancestor, current_vote_value + *votes);
+                total_vote_count += votes;
+            }
+        }
+        // Check if there is a clear block winner at this height. If so return it.
+        for (hash, votes) in current_votes.iter() {
+            if *votes > total_vote_count / 2 {
+                // we have a clear winner, return it
+                return Some(*hash);
+            }
+        }
+        // didn't find a clear winner
+        None
+    }
+
+    // Finds the best child (one with highest votes)
+    fn choose_best_child(&self, votes: &HashMap<Hash256, u64>) -> Option<Hash256> {
+        if votes.is_empty() {
+            return None;
+        }
+        let mut best_child: Hash256 = Hash256::from(0);
+        let mut max_votes: u64 = 0;
+        for (&candidate, &votes) in votes.iter() {
```

Comment:
@AgeManning made the changes. 
However, in case that all children have zero votes, it seems that the `votes` hashmap would be empty in the `choose_best_child` function. 
When I tried out with the following test case, the code panics with `CannotFindBestChild` because  `child_votes` hashmap never got populated in the first place and hence `choose_best_child` function returns None. 

- blocks:
    - id: 'b0'
      parent: 'b0'
    - id: 'b1'
      parent: 'b0'
    - id: 'b2'
      parent: 'b0'
  weights:
    - b1: 0
    - b2: 0
  heads:
    - id: 'b1'

I'm not sure what behaviour is expected here.

Reply:
Yep. This is doesn't conform with the phase 0 spec. I will correct both implementations. 
Thanks for pointing this out :)